File: src/main/java/com/flexpoker/game/command/eventpublishers/InMemoryAsyncGameEventPublisher.java
package com.flexpoker.game.command.eventpublishers;

import java.util.Timer;
import java.util.TimerTask;

import javax.inject.Inject;

import org.springframework.stereotype.Component;

import com.flexpoker.framework.event.EventPublisher;
import com.flexpoker.framework.event.subscriber.EventSubscriber;
import com.flexpoker.framework.processmanager.ProcessManager;
import com.flexpoker.game.command.events.GameStartedEvent;
import com.flexpoker.game.command.events.GameTablesCreatedAndPlayersAssociatedEvent;
import com.flexpoker.game.command.events.NewHandIsClearedToStartEvent;
import com.flexpoker.game.command.events.PlayerMovedToNewTableEvent;
import com.flexpoker.game.command.events.TablePausedForBalancingEvent;
import com.flexpoker.game.command.events.TableResumedAfterBalancingEvent;
import com.flexpoker.game.command.framework.GameEvent;

@Component
public class InMemoryAsyncGameEventPublisher implements EventPublisher<GameEvent> {

    private final EventSubscriber<GameEvent> gameEventSubscriber;

    private final ProcessManager<GameTablesCreatedAndPlayersAssociatedEvent> createInitialTablesForGameProcessManager;

    private final ProcessManager<GameStartedEvent> startFirstHandProcessManager;

    private final ProcessManager<NewHandIsClearedToStartEvent> startNewHandForExistingTableProcessManager;

    private final ProcessManager<GameStartedEvent> incrementBlindsCountdownProcessManager;

    private final ProcessManager<PlayerMovedToNewTableEvent> movePlayerBetweenTablesProcessManager;

    private final ProcessManager<TablePausedForBalancingEvent> pauseTableProcessManager;

    private final ProcessManager<TableResumedAfterBalancingEvent> resumeTableProcessManager;

    @Inject
    public InMemoryAsyncGameEventPublisher(
            EventSubscriber<GameEvent> gameEventSubscriber,
            ProcessManager<GameTablesCreatedAndPlayersAssociatedEvent> createInitialTablesForGameProcessManager,
            ProcessManager<GameStartedEvent> startFirstHandProcessManager,
            ProcessManager<NewHandIsClearedToStartEvent> startNewHandForExistingTableProcessManager,
            ProcessManager<GameStartedEvent> incrementBlindsCountdownProcessManager,
            ProcessManager<PlayerMovedToNewTableEvent> movePlayerBetweenTablesProcessManager,
            ProcessManager<TablePausedForBalancingEvent> pauseTableProcessManager,
            ProcessManager<TableResumedAfterBalancingEvent> resumeTableProcessManager) {
        this.gameEventSubscriber = gameEventSubscriber;
        this.createInitialTablesForGameProcessManager = createInitialTablesForGameProcessManager;
        this.startFirstHandProcessManager = startFirstHandProcessManager;
        this.startNewHandForExistingTableProcessManager = startNewHandForExistingTableProcessManager;
        this.incrementBlindsCountdownProcessManager = incrementBlindsCountdownProcessManager;
        this.movePlayerBetweenTablesProcessManager = movePlayerBetweenTablesProcessManager;
        this.pauseTableProcessManager = pauseTableProcessManager;
        this.resumeTableProcessManager = resumeTableProcessManager;
    }

    @Override
    public void publish(GameEvent event) {
        gameEventSubscriber.receive(event);

        if (event
                .getClass() == GameTablesCreatedAndPlayersAssociatedEvent.class) {
            createInitialTablesForGameProcessManager
                    .handle((GameTablesCreatedAndPlayersAssociatedEvent) event);
        } else if (event.getClass() == GameStartedEvent.class) {
            final var localStartFirstHandProcessManager = this.startFirstHandProcessManager;
            final var timer = new Timer();
            final var timerTask = new TimerTask() {
                @Override
                public void run() {
                    localStartFirstHandProcessManager.handle((GameStartedEvent) event);
                }
            };
            timer.schedule(timerTask, 10000);

            incrementBlindsCountdownProcessManager.handle((GameStartedEvent) event);
        } else if (event.getClass() == NewHandIsClearedToStartEvent.class) {
            startNewHandForExistingTableProcessManager
                    .handle((NewHandIsClearedToStartEvent) event);
        } else if (event.getClass() == PlayerMovedToNewTableEvent.class) {
            movePlayerBetweenTablesProcessManager
                    .handle((PlayerMovedToNewTableEvent) event);
        } else if (event.getClass() == TablePausedForBalancingEvent.class) {
            pauseTableProcessManager
                    .handle((TablePausedForBalancingEvent) event);
        } else if (event.getClass() == TableResumedAfterBalancingEvent.class) {
            resumeTableProcessManager
                    .handle((TableResumedAfterBalancingEvent) event);
        }
    }
}

File: src/main/kotlin/com/flexpoker/game/command/eventpublishers/InMemoryAsyncGameEventPublisher.kt
package com.flexpoker.game.command.eventpublishers

import com.flexpoker.framework.event.EventHandler
import com.flexpoker.framework.event.EventPublisher
import com.flexpoker.framework.event.subscriber.EventSubscriber
import com.flexpoker.framework.processmanager.ProcessManager
import com.flexpoker.game.command.events.GameEvent
import com.flexpoker.game.command.events.GameStartedEvent
import com.flexpoker.game.command.events.GameTablesCreatedAndPlayersAssociatedEvent
import com.flexpoker.game.command.events.NewHandIsClearedToStartEvent
import com.flexpoker.game.command.events.PlayerMovedToNewTableEvent
import com.flexpoker.game.command.events.TablePausedForBalancingEvent
import com.flexpoker.game.command.events.TableResumedAfterBalancingEvent
import org.springframework.stereotype.Component
import java.util.Timer
import java.util.TimerTask
import javax.inject.Inject

@Component
class InMemoryAsyncGameEventPublisher @Inject constructor(
    private val gameEventSubscriber: EventSubscriber<GameEvent>,
    private val createInitialTablesForGameProcessManager: ProcessManager<GameTablesCreatedAndPlayersAssociatedEvent>,
    private val startFirstHandProcessManager: ProcessManager<GameStartedEvent>,
    private val startNewHandForExistingTableProcessManager: ProcessManager<NewHandIsClearedToStartEvent>,
    private val incrementBlindsCountdownProcessManager: ProcessManager<GameStartedEvent>,
    private val movePlayerBetweenTablesProcessManager: ProcessManager<PlayerMovedToNewTableEvent>,
    private val pauseTableProcessManager: ProcessManager<TablePausedForBalancingEvent>,
    private val resumeTableProcessManager: ProcessManager<TableResumedAfterBalancingEvent>
) : EventPublisher<GameEvent> {

    companion object {
        private val NOOP = EventHandler { _: GameEvent -> }
    }

    override fun publish(event: GameEvent) {
        gameEventSubscriber.receive(event)
        when (event) {
            is GameTablesCreatedAndPlayersAssociatedEvent -> createInitialTablesForGameProcessManager.handle(event)
            is GameStartedEvent -> {
                val localStartFirstHandProcessManager = startFirstHandProcessManager
                val timer = Timer()
                val timerTask: TimerTask = object : TimerTask() {
                    override fun run() {
                        localStartFirstHandProcessManager.handle(event)
                    }
                }
                timer.schedule(timerTask, 10000)
                incrementBlindsCountdownProcessManager.handle(event)
            }
            is NewHandIsClearedToStartEvent -> startNewHandForExistingTableProcessManager.handle(event)
            is PlayerMovedToNewTableEvent -> movePlayerBetweenTablesProcessManager.handle(event)
            is TablePausedForBalancingEvent -> pauseTableProcessManager.handle(event)
            is TableResumedAfterBalancingEvent -> resumeTableProcessManager.handle(event)
            else -> NOOP
        }
    }

}